pkg/pillar/tgt/tgt.go

```diff
@@ -0,0 +1,168 @@
+// Copyright (c) 2021 Zededa, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package tgt
+
+import (
+	"fmt"
+	"io/ioutil"
+	"math/rand"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	tgtPath    = "/hostfs/sys/kernel/config/target"
+	iBlockPath = tgtPath + "/core/iblock_0"
+	naaPrefix  = "5001405" // from rtslib-fb
+)
+
+func waitForFile(fileName string) error {
+	maxDelay := time.Second * 5
+	delay := time.Millisecond * 500
+	var waited time.Duration
+	for {
+		if delay != 0 {
+			time.Sleep(delay)
+			waited += delay
+		}
+		if _, err := os.Stat(fileName); err == nil {
+			return nil
+		} else {
+			if waited > maxDelay {
+				return fmt.Errorf("file not found: error %v", err)
+			}
+			delay = 2 * delay
+			if delay > maxDelay {
+				delay = maxDelay
+			}
+		}
+	}
+}
+
+//CheckTargetIBlock check target iblock exists
+func CheckTargetIBlock(tgtName string) bool {
+	targetRoot := filepath.Join(iBlockPath, tgtName)
+	if _, err := os.Stat(targetRoot); err == nil {
+		return true
+	}
+	return false
+}
+
+// TargetCreateIBlock - Create iblock target for device
+func TargetCreateIBlock(dev, tgtName, serial string) error {
+	targetRoot := filepath.Join(iBlockPath, tgtName)
+	err := os.MkdirAll(targetRoot, os.ModeDir)
+	if err != nil {
+		return fmt.Errorf("cannot create fileio: %v", err)
+	}
+	if err := waitForFile(filepath.Join(targetRoot, "control")); err != nil {
+		return fmt.Errorf("error waitForFile: %v", err)
+	}
+	controlCommand := fmt.Sprintf("udev_path=%s", dev)
+	if err := ioutil.WriteFile(filepath.Join(targetRoot, "control"), []byte(controlCommand), 0660); err != nil {
+		return fmt.Errorf("error set control: %v", err)
+	}
+	if err := ioutil.WriteFile(filepath.Join(targetRoot, "wwn", "vpd_unit_serial"), []byte(serial), 0660); err != nil {
+		return fmt.Errorf("error set vpd_unit_serial: %v", err)
+	}
+	if err := ioutil.WriteFile(filepath.Join(targetRoot, "enable"), []byte("1"), 0660); err != nil {
+		return fmt.Errorf("error set enable: %v", err)
+	}
+	return nil
+}
+
+//GenerateNaaSerial generates naa serial
+func GenerateNaaSerial() string {
+	rand.Seed(time.Now().UnixNano())
```

Comment:
I don't know what clock resolution we might have on our devices; supposed it is a millisecond then we might see two created with the same seed.

Two possibilities: 1. Seed in a Init function. 2. Pass in a unique name (the volume uuid?) and add that to the seed.

Reply:
I moved it to init, thanks for suggestions